imagenameparser/api/image_name.py

```diff
@@ -0,0 +1,872 @@
+# Copyright (c) 2022 SUSE LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+# <http://www.apache.org/licenses/LICENSE-2.0>
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+"""Provides the ``ImageName`` API class"""
+
+import logging
+
+from imagenameparser.api.image_name_regexp import ImageNameRegexp
+from imagenameparser.errors import BadRegexMatchError, InvalidFrameworkError
+
+_logger = logging.getLogger(__name__)
+
+
+class ImageName:  # pylint: disable=R0904
+    """
+    Parse an image name.
+
+    ``framework`` should be a valid cloud provider framework name, such
+    as ``ec2`` or ``azure``.
+
+    ``name`` should be a valid image name, such as any of the image names
+    reported by ``pint``.
+
+    """
+
+    def __init__(self, framework, name):
+        self._framework = framework
+        self._image_name = name
+        self._image_info = None
+
+        self._check_framework()
+        self._extract_image_info()
+
+    def _check_framework(self):
+        """Check if framework is valid cloud provider framework name
+
+        Raises:
+            FrameworkError: If the framework name is missing or Invalid
+        """
+
+        if self._framework is None:
+            raise InvalidFrameworkError("framework is required", "")
+        if self._framework not in (
+            'ec2',
+            'azure',
+            'gce',
+            'oci',
+            'aliyun',
+            'amazon',
+            'alibaba',
+            'oracle',
+            'google',
+                'microsoft'):
```

Comment:
indentation

Something is not quite right here. Maybe @KeithMnemonic or @rtamalin can weigh in on this as well. The tests tests/test_imagenameparser.py clearly use permutations, names such as `amazon` and `microsoft` but I am missing the mapping implementation, i.e. when the user specifies `amazon` it should be mapped to `ec2` otherwise the test on line 81 is incorrect and the `is_ec2()` test on line 104 will also be incorrect.

I was under the impression that so far we had delegated the mapping to the client and that within the name handling we only supported

`aliyun,azure,ec2, gce, oci`

If we allow `alibaba`, `amazon`, `microsoft` etc. then all the tests if `is_*()` have to be adjusted accordingly or we need to update the assignment of `self._framework` accordingly.

Something is not quite right here.

Reply:
Since this is a new project it would be nice, if possible, to fix the naming inconsistencies.  At some point we agreed to names of providers and not frameworks or acronyms. E.g. names would be:

- amazon
- alibaba
- google
- microsoft

Reply:
@smarlowucf  Did you mean change mashouse as well. Currently the fragments use the framerwork and not the providers.  Or did you mean change the cli/api and have an internal mapping from providers/frameworks.  @rjschwei I agree about the tests as the mashhouse cli only accepts `aliyun,azure,ec2, gce, oci` I  think these are included.
```
            'amazon',
            'alibaba',
            'oracle',
            'google',
            'microsoft'
```

due to the pint UI being able to use this API and it only has providers and not frameworks.  @jeremy-moffitt can speak about the pint UI use case.   So perhaps it would be best to be consistent and switch to providers and then patch mashhouse once it can also use this API?

Reply:
I think there is general agreement that the current implementation is broken and as such should not be merged. My suggestion for the path forward is to have the initial commit with `aliyun,azure,ec2, gce, oci` as accepted names only. Then we can have a follow up conversation as to whether we want to make code changes such that `amazon, alibaba, oracle, google, microsoft` are accepted in addition or instead of the existing set.

Reply:
@KeithMnemonic yeah with newer tools we've been using provider names only. I plan to cleanup mashhouse and all other tools eventually. But it's of course lower priority to clean up existing naming. Thus it will take time.